File: Assets/src/model/grasses/DeathlyCreeper.cs
using System.Linq;
using UnityEngine;

[System.Serializable]
[ObjectInfo(description: "Spreads to unoccupied adjacent Tiles without Grass.\n\nIf all Ground tiles have Black Creeper, all creatures (including you) Die.")]
public class DeathlyCreeper : Grass, /*IActorEnterHandler,*/ ISteppable {
  public override string displayName => "Black Creeper";
  public static bool CanOccupy(Tile tile) => tile.CanBeOccupied() && tile is Ground && !(tile.grass is DeathlyCreeper) /*&& tile.grass == null*/;
  public DeathlyCreeper(Vector2Int pos) : base(pos) {
    timeNextAction = timeCreated + 1;
  }

  static float timeLastChecked = -1f;

  public float timeNextAction { get; set; }

  public float turnPriority => 50;

  public float Step() {
    if (timeLastChecked != GameModel.main.time) {
      CheckWin();
    }

    // if (actor != null) {
    //   OnNoteworthyAction();
    //   actor.TakeDamage(1, this);
    // }
    if (age % 4 == 3) {
      var adjacent = Util.RandomPick(floor.GetCardinalNeighbors(pos).Where(CanOccupy));
      if (adjacent != null) {
        OnNoteworthyAction();
        floor.Put(new DeathlyCreeper(adjacent.pos));
      }
    }
    return 1;
  }

  private void CheckWin() {
    bool isLevelCovered = floor.tiles.OfType<Ground>().All(tile => tile.grass is DeathlyCreeper);
    if (isLevelCovered) {
      GameModel.main.EnqueueEvent(() => {
        // copy body list in case it's modified
        foreach (var body in floor.bodies.ToList()) {
          body.Kill(this);
        }
      });
    }
  }
  // public void HandleActorEnter(Actor who) {
  //   who.TakeDamage(1, this);
  //   // who.Kill(this);
  //   Kill(who);
  // }
}

File: web/src/model/grasses/DeathlyCreeper.ts
import { Grass } from './Grass';
import { Ground } from '../Tile';
import { GameModelRef } from '../GameModelRef';
import { MyRandom } from '../../core/MyRandom';
import { entityRegistry } from '../../generator/entityRegistry';
import type { ISteppable } from '../Floor';
import type { Vector2Int } from '../../core/Vector2Int';
import type { Tile } from '../Tile';

let timeLastChecked = -1;

/**
 * Spreads to adjacent Ground tiles every 4 turns.
 * If all Ground tiles are covered, kills all bodies.
 * Port of C# DeathlyCreeper.cs.
 */
export class DeathlyCreeper extends Grass implements ISteppable {
  timeNextAction: number;
  get turnPriority(): number { return 50; }

  get displayName(): string { return 'Black Creeper'; }

  static canOccupy(tile: Tile): boolean {
    return tile.canBeOccupied() && tile instanceof Ground && !(tile.grass instanceof DeathlyCreeper);
  }

  constructor(pos: Vector2Int) {
    super(pos);
    this.timeNextAction = this.timeCreated + 1;
  }

  step(): number {
    if (timeLastChecked !== GameModelRef.main.time) {
      this.checkWin();
    }

    if (this.age % 4 === 3) {
      const neighbors = this.floor!.getCardinalNeighbors(this.pos)
        .filter(t => DeathlyCreeper.canOccupy(t));
      const target = MyRandom.Pick(neighbors);
      if (target) {
        this.onNoteworthyAction();
        this.floor!.put(new DeathlyCreeper(target.pos));
      }
    }
    return 1;
  }

  private checkWin(): void {
    timeLastChecked = GameModelRef.main.time;
    // Check if all Ground tiles are covered by DeathlyCreeper
    let allCovered = true;
    for (const tile of this.floor!.enumerateFloor()) {
      const t = this.floor!.tiles.get(tile);
      if (t instanceof Ground && !(t.grass instanceof DeathlyCreeper)) {
        allCovered = false;
        break;
      }
    }
    if (allCovered) {
      GameModelRef.main.enqueuEvent(() => {
        const bodies = [...this.floor!.bodies];
        for (const body of bodies) {
          body.kill(this);
        }
      });
    }
  }
}

entityRegistry.register('DeathlyCreeper', DeathlyCreeper);
